### PR TITLE
Extend JavaDocs to remove errors, compatibility with Pipelines

### PR DESCRIPTION
**Pull request by zmeggyesi**

### CODE REVIEW

**Comment by zmeggyesi** [comment]
@Vlatombe
I'm not sure this plugin is even maintained any more. Is it? :)

**Comment by zmeggyesi** [comment]
@ndeloof 
Sorry to bother you, but do you think this plugin is being maintained? I have a fix that enables compatibility with the scripted pipelines that could be merged and a new version released.

**Comment by ndeloof** [comment]
@zmeggyesi can't tell, join jenkins-dev mailing list to ask.

File: pom.xml [hunk 1]
```diff
@@ -4,14 +4,18 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.532.3</version>
+    <version>2.31-SNAPSHOT</version>
```
Comment (slide) [inline pom.xml]: Should not depend on SNAPSHOT releases

File: pom.xml [hunk 2]
```diff
@@ -4,14 +4,18 @@
...
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.532.3</version>
+    <version>2.31-SNAPSHOT</version>
   </parent>
   <artifactId>validating-string-parameter</artifactId>
   <packaging>hpi</packaging>
   <name>Validating String Parameter Plugin</name>
-  <version>2.4-SNAPSHOT</version>
+  <version>2.5-SNAPSHOT</version>
```
Comment (slide) [inline pom.xml]: Is there a reason the version needed to be bumped here? Usually the pom is left in a place that is correct for the next round of development.

File: pom.xml [hunk 3]
```diff
@@ -4,14 +4,18 @@
...
   <url>http://wiki.jenkins-ci.org/display/JENKINS/Validating+String+Parameter+Plugin</url>
 
+  <properties>
+    <jenkins.version>2.46.2</jenkins.version>
```
Comment (slide) [inline pom.xml]: This is most likely not correct. I recommend not providing this property and allowing the default for the parent pom to be used. This would mean a whole lot of users still on 1.x would not be able to use this plugin.

Comment (zmeggyesi) [inline pom.xml]: Definitely good point. I had to up this so that `hpi:run` started the correct version of Jenkins, and forgot to restore it afterwards. Did that now.

File: pom.xml (hunk 1 above)
Comment (zmeggyesi) [inline pom.xml]: True, I wanted to check compatibility going forward. That said, would you be all right with depending on the latest stable version (2.32)?

File: pom.xml (hunk 2 above)
Comment (zmeggyesi) [inline pom.xml]: My reasoning was that in keeping with semantic versioning, the minor number should be increased: this is a feature of the plugin that did not exist previously, but does not introduce backwards-incompatible changes. Thus, it doesn't warrant a major version bump, but certainly warrants one.

Since there are no clear-cut contributor guides, I'll go with what you decide is convention.

File: pom.xml (hunk 1 above)
Comment (slide) [inline pom.xml]: For the plugin parent version, that should be fine.

**Comment by slide** [comment]
It's very hard to tell what you changed in your latest commits because the whole pom.xml shows as changed. Did your editor change the whitespace?

File: pom.xml (hunk 2 above)
Comment (slide) [inline pom.xml]: When someone does a release of the plugin, the pom.xml gets updated for the NEXT version automatically, so you don't need to change the version.

**Comment by zmeggyesi** [comment]
It shouldn't have, but since I use two different dev machines, the one I used now might have had a misconfigured line ending setting.

The only change was supposed to be line 7: I removed the SNAPSHOT version in favor of 2.32; and remove the Jenkins version requirement.
I'll reset the version value, and leave it to your automation, now that I know how it works.

File: pom.xml [hunk 4]
```diff
@@ -1,56 +1,53 @@
...
-      <plugin>
-        <artifactId>maven-release-plugin</artifactId>
-        <version>2.5.1</version>
-      </plugin>
-    </plugins>
-  </build>
-</project>  
-
-
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
```
Comment (oleg-nenashev) [inline pom.xml]: 🐜 unrelated formatting changes

File: pom.xml [hunk 5]
```diff
@@ -1,56 +1,53 @@
...
+      <name>Peter Hayes</name>
+      <email>petehayes@gmail.com</email>
+      <timezone>-6</timezone>
+    </developer>
+  </developers>
+
+  <repositories>
+    <repository>
+      <id>repo.jenkins-ci.org</id>
+      <url>http://repo.jenkins-ci.org/public/</url>
```
Comment (oleg-nenashev) [inline pom.xml]: Use HTTPS then

File: src/main/java/hudson/plugins/validating_string_parameter/ValidatingStringParameterDefinition.java [hunk 6]
```diff
@@ -118,6 +119,11 @@ public FormValidation doCheckRegex(@QueryParameter final String value) {
 
         /**
          * Called to validate the passed user entered value against the configured regular expression.
+         *
+         * @param regex the regex
+         * @param failedValidationMessage message to display when validation fails
+         * @param value the value to be validated
+         * @return whether the value is valid according to the specified regex
          */
         public FormValidation doValidate(@QueryParameter("regex") String regex,
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/validating_string_parameter/ValidatingStringParameterDefinition.java]: Nice to have `@Restricted(NoExternalUse.class)`. It prevents usage of Stapler methods in dependent plugins

File: src/main/java/hudson/plugins/validating_string_parameter/ValidatingStringParameterValue.java [hunk 7]
```diff
@@ -73,13 +76,27 @@ public BuildWrapper createBuildWrapper(AbstractBuild<?, ?> build) {
...
+                    throw new AbortException("Invalid value for parameter [" + getName() + "] specified: " + value);
+                }
             };
         } else {
             return null;
         }
     }
 
     @Override
+    public void buildEnvironment(Run<?, ?> build, EnvVars env) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/validating_string_parameter/ValidatingStringParameterValue.java]: It does not seem to be the best approach. Maybe makes sense to use `QueueDecisionHandler` or something like that

File: pom.xml (hunk 5 above)
Comment (zmeggyesi) [inline pom.xml]: Didn't touch this part, but sure.

File: src/main/java/hudson/plugins/validating_string_parameter/ValidatingStringParameterValue.java (hunk 7 above)
Comment (zmeggyesi) [inline src/main/java/hudson/plugins/validating_string_parameter/ValidatingStringParameterValue.java]: I'm not sure I'm seeing how: as long as the plugin is extending `ParameterValue`, I'm not seeing where I could pull this class from, and frankly, I'm not sure I have the experience in plugins to do the rewrite necessary to use the `QueueDecisionHandler`...

**Comment by slide** [comment]
Superseded by #11
